### PR TITLE
78 update car

### LINKED ISSUE #78: [back] update car

### PR DESCRIPTION
**Pull request by husamkamal**

### CODE REVIEW

File: client/src/pages/Cars/index.tsx [hunk 1]
```diff
@@ -17,6 +17,7 @@ function Cars() {
     setCurrentPAge(value);
     window.scrollTo({ top: 0, behavior: 'smooth' });
   };
+  console.log(cars, 'cars');
```
Comment (LinaYahya) [inline client/src/pages/Cars/index.tsx]:
```suggestion
```

File: .github/workflows/node.js.yml [hunk 2]
```diff
@@ -53,3 +53,4 @@ jobs:
 
         env: 
           DB_URL_TEST: postgres://team3:12345@localhost:5432/goodcar
+          SECRET_KEY: car-trade-secret-123456
```
Comment (LinaYahya) [inline .github/workflows/node.js.yml]: I'll put both secret key and tokens as repository secrets,
Check How to import them [encrypted secret](https://docs.github.com/en/actions/security-guides/encrypted-secrets)

File: server/__tests__/cars.ts [hunk 3]
```diff
@@ -92,4 +92,12 @@ describe('/cars endpoint', () => {
     expect(result.body.data.rows[0].price).toBeLessThanOrEqual(200000);
     expect(result.statusCode).toEqual(200);
   });
+  test('should return id of car that update', async () => {
+    const result = await request(app).put('/api/v1/cars/1')
+
+      .set('Cookie', 'token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyb2xlIjoiYWRtaW4iLCJ1c2VySWQiOjF9.5bbTDHhlnCvs5ZDGfXcL_MU_4CCKzJ61wUDCGD_nWks');
```
Comment (LinaYahya) [inline server/__tests__/cars.ts]: put on env and import from there, 
Also I've put it as a repo secret as `ADMIN_TOKEN`

File: server/src/controllers/cars/updateCars.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+import createError from 'http-errors';
+import { updateCarServes } from '../../services/cars';
+import updateCarSchema from '../../helpers/updateCarSchema';
+
+const updateCars = async (req: Request, res: Response, next:NextFunction) => {
+  const { body } = req;
+  const { id } = req.params;
+  updateCarSchema.validate(body).catch((err) => {
+    next(createError(404, err.errors));
+  });
```
Comment (LinaYahya) [inline server/src/controllers/cars/updateCars.ts]: Why you do this, if there as an error it'll be thrown to `RouterWraper` where we handle validation error

File: server/src/helpers/updateCarSchema.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
+import * as yup from 'yup';
```
Comment (LinaYahya) [inline server/src/helpers/updateCarSchema.ts]: We already have a validation folder

File: server/__tests__/cars.ts [hunk 6]
```diff
@@ -92,4 +92,12 @@ describe('/cars endpoint', () => {
     expect(result.body.data.rows[0].price).toBeLessThanOrEqual(200000);
     expect(result.statusCode).toEqual(200);
   });
+  test('should return id of car that update', async () => {
```
Comment (LinaYahya) [inline server/__tests__/cars.ts:97]: create more tests in case of token was malformed, or where is no token

File: server/__tests__/cars.ts [hunk 7]
```diff
@@ -92,4 +92,12 @@ describe('/cars endpoint', () => {
...
+    const result = await request(app).put('/api/v1/cars/1')
+
+      .set('Cookie', 'token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyb2xlIjoiYWRtaW4iLCJ1c2VySWQiOjF9.5bbTDHhlnCvs5ZDGfXcL_MU_4CCKzJ61wUDCGD_nWks');
+    expect(result.body.data[0]).toEqual(0);
+    expect(result.body.msg).toEqual('done!');
+    expect(result.statusCode).toEqual(200);
```
Comment (LinaYahya) [inline server/__tests__/cars.ts:102]: Is that a real test!!
Where's request body?

File: server/src/helpers/updateCarSchema.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+const updateCarSchema = yup.object().shape({
+
```
Comment (LinaYahya) [inline server/src/helpers/updateCarSchema.ts]: What about state and fuel, validate them according to options we have

File: server/src/services/cars/updateCar.ts [hunk 9]
```diff
@@ -0,0 +1,12 @@
+import { Car } from '../../db/models';
+
+const updateCarServes = async (body, id) => {
+  const car = await Car.update(
+    { ...body },
+    { where: { id } },
+  );
+  console.log(car);
```
Comment (LinaYahya) [inline server/src/services/cars/updateCar.ts]:
```suggestion
```

File: server/src/controllers/cars/updateCars.ts [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+  const result = await updateCarServes(body, id);
+  return { status: 200, msg: 'done!', data: result };
```
Comment (LinaYahya) [inline server/src/controllers/cars/updateCars.ts:11]:
```suggestion
  return { status: 200, msg: 'done!', data: result }; // not sure if we want to return count of rows updated, No need I guess
```

**Review by LinaYahya — CHANGES_REQUESTED** [review]

**Review by abdou059 — APPROVED** [review]

File: server/src/validation/updateCarSchema.ts [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+  price: yup.number().notRequired().label('price'),
+  year: yup.number().notRequired().label('year'),
+  mileage: yup.number().notRequired().label('mileage'),
+  quality: yup.number().notRequired().label('quality'),
+  isGoodPrice: yup.boolean().notRequired().label('isGoodPrice'),
+  features: yup.array().of(yup.string()).notRequired().label('features'),
+  transmission: yup.string().notRequired().label('transmission'),
+  description: yup.string().notRequired().label('description'),
+  fuel: yup.string().notRequired().label('fuel'),
+  state: yup.string().notRequired().label('state'),
```
Comment (LinaYahya) [inline server/src/validation/updateCarSchema.ts]: state here accept any value is it's only required, But we need to validate value to be one the options allowed,
Check `Yup.oneof()`,
The same for fuel

File: server/src/controllers/cars/updateCars.ts (hunk 10 above)
Comment (LinaYahya) [reply server/src/controllers/cars/updateCars.ts:11]: Not sure if it is resolved!

**Review by LinaYahya — CHANGES_REQUESTED** [review]

File: server/src/validation/updateCarSchema.ts [hunk 12]
```diff
@@ -0,0 +1,19 @@
+import * as yup from 'yup';
```
Comment (LinaYahya) [inline server/src/validation/updateCarSchema.ts:1]: export to `validation/index` and import from there

**Review by LinaYahya — CHANGES_REQUESTED** [review]

**Review by LinaYahya — APPROVED** [review]

**Review by mai-mohammed — APPROVED** [review]

**Pull request merged**